metrics/tagindex.go

```diff
@@ -0,0 +1,132 @@
+package metrics
+
+import (
+	"sync"
+
+	"github.com/mstoykov/atlas"
+)
+
+// FIXME: naming! we now have TagMap, SampleTags, SystemTag, TagSet
+// so the naming decision and maybe some sort of uniquification could be good.
+//
+// TagSet should be the best but is already used, TagIndex could be a valid alternative?
+
+// TagMap represents a set of tags.
+type TagMap struct {
+	mutex sync.RWMutex
+	tags  *atlas.Node
+	locks bool
```

Comment:
Can we get some comment explaining why this exists?

Reply:
It's a wrapper for the tag set concept around Atlas, essentially hiding the `n = n.AddLink()` logic behind an `AddTag` method, From/To SampleTags and the concurrency handling when required. It is the same as SampleTags but mutable.

The concurrency feature it's mostly required for removing the TagMap from [the lib](https://github.com/grafana/k6/blob/521a2823afd0018cb8d9f8a2931c18dd63ad0010/lib/state.go#L97-L101) package and use a unique struct.

So a comment could be like the following:

```suggestion
	// TagMap represents a ordered set of tags, it's expected to be initialized once
	// with an empty or a root set of tags and use BranchOut for generating from it all the new tag set,
	// in this way, TagMap can optimize and reduce as much as possible the memory allocation.
        // By default a TagMap is not concurrent safe, check the `SetSafeConcurrency` method to get it.
        type TagMap struct {
	  mutex sync.RWMutex
	  tags  *atlas.Node
	  locks bool
```

Reply:
I am not sure which is better, this property or separate types :thinking:  Let me see if I understand things correctly:
1. we don't have to account for any concurrency when building the atlas `Node` that has `--tags` (the de-facto root of the test run)
2. we also don't have any concurrency when setting the `Node` that corresponds to the initial `scenario.tags` in every VU
3. we need thread safety when dealing with VU tags during execution, both because `lib.State.Tags` can be changed from `k6/execution`, and because we can have extra metric tags everywhere (custom metrics, HTTP requests, etc.), however:
    1. we only need to ensure thread safety when we access the VU tags, once we branch out of them, we don't need any locking
    2. the actual building of the tags (e.g. for a single HTTP request) is sequential, tag by tag
   
Am I missing something? Because, to me, it seems like we need only very little locking in the few places that read and write the current VU tags in `lib.State`? I think everything else besides that can just depend on atlas' internal locking? :thinking:

Reply:
>Am I missing something? Because, to me, it seems like we need only very little locking in the few places that read and write the current VU tags in lib.State? I think everything else besides that can just depend on atlas' internal locking?

Your description is right, there aren't missing points :+1: 

My rationale behind it: I thought it could increase the complexity of our project structure to have one more struct doing the same thing (+ locks). I found moving the locks on `metrics.TagMap` and removing the `lib.TagMap` a convenient simplification of our Tag representations. I thought also that it could be useful in the future, now that we have the event loop and concurrent operations within iterations (but I don't have a concrete use-case to point for that).

Reply:
That's the thing though - you might have concurrent operations, you can kind of have them even now (e.g. with `http.batch()` or websockets), but when we are measuring something and building the tag set, the only part where we need to do locking is getting the current VU's tags. After we get them, the building of the tag set for the current measurement is actually sequential. 

For example, we have already done the HTTP request, and we have already measured it, now we are just transfering these observations and measurements to some other object, but we do this perfectly sequentially and without any need for locking (besides the atlas internals, just because that's how it works). That's why we don't have any locking here now: https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/lib/netext/httpext/transport.go#L95

Reply:
>but when we are measuring something and building the tag set, the only part where we need to do locking is getting the current VU's tags. After we get them, the building of the tag set for the current measurement is actually sequential.

This is how the `metrics.TagMap` implementation works, the default behaviour doesn't use locks, it isn't concurrent safe. `SetConcurrentSafe` has to be invoked if safety is required and in the current implementation is invoked **only** for the `lib.State.Tags`.

Reply:
If this is creating confusion we can split them, but one will be the replica of the other + locks.

Reply:
> If this is creating confusion we can split them, but one will be the replica of the other + locks.

I'd prefer that approach - it's a fairly small data structure. 

I think the type with locking will only be needed in a single place, in the [VU's `lib.State.Tags`](https://github.com/grafana/k6/blob/1e8468292ab5eb0d2333b1278b44e81c5cbef679/lib/state.go#L76), right? And besides, it won't be exactly the same type - keep in mind we won't just need to lock the current atlas `Node` in the VU when we want to build and emit a `Sample`, we'll also need to lock the unindexable tags (https://github.com/grafana/k6/issues/2584) that are currently set in the VU's `State`. In my now hopelessly conflicted and obsolete branch, I didn't have atlas, but I basically continued having `lib.State.Tags` to be its own type with internal locking, just extended its `Clone` method.